Have LoggerServiceFactory inject writer and processor plugin managers

TotalWipeOut opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no

### Description

The issue: `LoggerServiceFactory` does not inject the writer nor the processor plugin managers. But `LoggerAbstractServiceFactory` does. And I want to use the former with custom processors.

I changed the inheritance so `LoggerAbstractServiceFactory` extends `LoggerServiceFactory` and moved some logic, so both inject the relevant plugin managers in the same way.

I changed `PsrLoggerAbstractAdapterFactory` to compose  `LoggerAbstractServiceFactory`, not inherit. It didn't make sense as the return types for `__invoke` were incompatible. So adding strict typing across much of the factories would be possible now

Added unit tests for `LoggerServiceFactory`. All other unit tests still pass, and the factories can still be used in exactly the same way. Only difference is now `LoggerServiceFactory` injects the plugin managers.

Typing only added to anything new to maintain BC

I want to target the 2.15.x branch, as i see this as a bugfix - however, i am new to open source, and am open to guidance.

Ocramius commented:
@TotalWipeOut probably need only a086d99, right?

Ocramius commented:
Tests are needed for this addition.

TotalWipeOut commented:
Thanks @Ocramius - yes only meant that one commit, i will redo this.

OK, i will write some tests for this change - thanks

Ocramius commented:
@TotalWipeOut you can force-push a cleaned up branch to your own PR branch BTW - no need to re-create this.

froschdesign commented on src/PsrLoggerAbstractAdapterFactory.php:
```diff
@@ -5,25 +5,80 @@
         return new PsrLoggerAdapter($logger);
     }
+
+    /**
+     * Determine if we can create a service with name
+     *
+     * @param string $name
+     * @param string $requestedName
+     */
+    public function canCreateServiceWithName(ServiceLocatorInterface $serviceLocator, $name, $requestedName): bool
+    {
+        return $this->loggerAbstractServiceFactory->canCreate($serviceLocator, $name);
+    }
```
@Ocramius 
laminas-log uses laminas-servicemanager 3 and methods like these are from version 2 of laminas-servicemanager. In other components this has always been removed, no matter if the factory class was final or not. Should the same approach be taken here?

Ocramius replied:
We should certainly push for v3 API.

I didn't check this yet, only skimmed and saw no tests, hence my initial feedback :)

TotalWipeOut replied:
Sure, I will remove v2 methods, very happy about that :+1:

TotalWipeOut committed:
LoggerFactory: inject plugin managers. PsrLoggerAbstractAdapterFactory now composes LoggerAbstractServiceFactory

Signed-off-by: Kevin Hamilton <kevin@eagleeye.com>

TotalWipeOut committed:
Tests for LoggerServiceFactory

Signed-off-by: Kevin Hamilton <kevin@eagleeye.com>

TotalWipeOut commented:
I have added tests for LoggerServiceFactory, the other factories were already fully tested.
I have removed all v2 API methods from factories too.
I am not sure why there are failing checks - they seem unrelated to my changes, but do say if there is an error somewhere. thanks

froschdesign commented:
@TotalWipeOut 
> I have removed all v2 API methods from factories too.

This also means to remove the interface `Interop\Container\ContainerInterface`.

TotalWipeOut commented:
@froschdesign 
> @TotalWipeOut
> 
> > I have removed all v2 API methods from factories too.
> 
> This also means to remove the interface `Interop\Container\ContainerInterface`.


I dont consider that v2 API, and because composer.json is set to `"laminas/laminas-servicemanager": "^3.3.0"`, and it was version `3.11.0` where the alias was added for Interop to Psr, to not BC break the interfaces

Changing Interop to Psr would require bumping version in composer.json, and i don't really want to change deps

Assuming i understood this correctly :)

froschdesign commented:
@TotalWipeOut 
laminas-servicemanager 3.3.0 uses container-interop/container-interop with version 1.2.0 and this means:

```php
namespace Interop\Container;

use Psr\Container\ContainerInterface as PsrContainerInterface;

interface ContainerInterface extends PsrContainerInterface
{
}
```

https://github.com/container-interop/container-interop/blob/79cbf1341c22ec75643d841642dd5d6acd83bdb8/src/Interop/Container/ContainerInterface.php#L13

> Changing Interop to Psr would require bumping version in composer.json, and i don't really want to change deps

Why not? After all, you are already working on this package, even though it is in [security-only maintenance mode](https://github.com/laminas/laminas-log#security-only-maintenance-mode).

TotalWipeOut commented:
@froschdesign 
https://github.com/laminas/laminas-servicemanager/blob/3.3.0/src/Factory/FactoryInterface.php#L11
The FactoryInterface requires `Interop`, the inheritance doesn't fix that. Using `Psr` namespace breaks it, try yourself using servicemanager 3.3.0 :)

The only way laminas was able to move to Psr, was via aliasing: https://github.com/laminas/laminas-servicemanager/blob/3.11.x/src/autoload.php

TotalWipeOut commented:
> Why not? After all, you are already working on this package

I just wanted to fix this bug, not refactor the package. Also, the current project i am working on is currently stuck on servicemanager 3.10.0 for other reasons, so doing this would push my fix out of my reach lol

froschdesign commented:
@TotalWipeOut 
My comment was misleading: you should not use version 3.3.0 of laminas-servicemanager, instead raise the version.

TotalWipeOut commented:
Ah, you know, I hadn't actually noticed that security only message :man_facepalming: 
So on those grounds, would this bugfix not be accepted...?

froschdesign commented:
> So on those grounds, would this bugfix not be accepted...?

All good, your changes will be accepted with thanks.

Ocramius commented on src/FilterPluginManagerFactory.php:
```diff
@@ -6,26 +6,22 @@
-     *
-     * @param array
-     */
-    protected $creationOptions;
```
Hmm, with this being a non-`final` class, this change is probably to be considered a BC break.

Same applies to signature changes

weierophinney replied:
Scheduling time to review tomorrow (2022-10-13)

TotalWipeOut replied:
thank you both for your time.

Happy to add back the v2 API stuff that i removed due to the comments above, if that avoids a BC break.

Ocramius reviewed (commented):
Letting @weierophinney review this, since my brain is fried today.

I think we have multiple BC breaks in the patch, right now.

weierophinney commented on src/FilterPluginManagerFactory.php:
```diff
@@ -6,26 +6,22 @@
+     * @throws NotFoundExceptionInterface
      */
-    public function __invoke(ContainerInterface $container, $name, ?array $options = null)
+    public function __invoke(ContainerInterface $container, $requestedName, ?array $options = null): FilterPluginManager
```
This constitutes a BC break, as any extensions of this class (and it **is** non-final) would break unless they also defined the same return type hint.

weierophinney commented on src/FormatterPluginManagerFactory.php:
```diff
@@ -6,27 +6,24 @@
+        ContainerInterface $container,
+        $requestedName,
+        ?array $options = null
+    ): FormatterPluginManager {
```
Same issue here with the return type hint.

weierophinney commented on src/LoggerServiceFactory.php:
```diff
@@ -16,27 +20,115 @@ class LoggerServiceFactory implements FactoryInterface
+     * @throws NotFoundExceptionInterface
      */
-    public function __invoke(ContainerInterface $container, $name, ?array $options = null)
+    public function __invoke(ContainerInterface $container, $requestedName, ?array $options = null): Logger
```
Again, return type hint change == BC break.

weierophinney commented on src/ProcessorPluginManagerFactory.php:
```diff
@@ -6,27 +6,23 @@
+        ContainerInterface $container,
+        $requestedName,
+        ?array $options = null
+    ): ProcessorPluginManager {
```
RTH change

weierophinney commented on src/FilterPluginManagerFactory.php:
```diff
@@ -6,26 +6,22 @@
-     *
-     * @param array
-     */
-    protected $creationOptions;
```
A note, in case somebody gets to this PR and wonders why these properties and methods are removed: It looks like we've not supported laminas-servicemanager v2 for some time now, so the v2 properties and methods are dead code.

(Likely should have called that out in the PR description; as it was, I had to verify this was the case during review.)

TotalWipeOut replied:
Understood. I will keep this change strictly to the bugfix

weierophinney commented on src/WriterPluginManagerFactory.php:
```diff
@@ -6,26 +6,19 @@
+     * @throws ContainerExceptionInterface
      */
-    public function __invoke(ContainerInterface $container, $name, ?array $options = null)
+    public function __invoke(ContainerInterface $container, $name, ?array $options = null): WriterPluginManager
```
RTH change

weierophinney reviewed (changes requested):
The **idea** behind the change is solid, and the way it is accomplished makes sense.

The problem I have with the PR is that it's doing a lot more than that:

- It's removing the laminas-servicemanager v2 properties and methods. Yes, these can be safely removed, but those changes are completely unrelated to what you're trying to accomplish. They should be done in a separate PR.
- It's adding return type hints to factories/abstract factories. Yes, this is good... but because these classes are not marked final, anybody who is extending and not defining an equivalent RTH will experience a BC break, since their own RTH will now not match (they'll be less restrictive, which breaks compatibility). Those changes should be backed out, or the PR should target next major.

TotalWipeOut commented:
@weierophinney thank you for the review. Understood. I will rollback the v2 changes and return types

TotalWipeOut committed:
Revoke return types to maintain BC, Remove used trait

Signed-off-by: Kevin Hamilton <kevin@eagleeye.com>

TotalWipeOut committed:
Add types to docblocks where missing

Signed-off-by: Kevin Hamilton <kevin@eagleeye.com>

TotalWipeOut commented:
@weierophinney I have, hopefully, addressed all your comments. I hope you can spare some time to check them out. Thanks in advance.

also, unsure why some of the github checks are failing, the reason shown i cannot replicate and doesn't seem related to my changes...

TotalWipeOut commented:
@weierophinney @Ocramius Would you have any time this week to finish off this PR? Thanks

Ocramius commented:
Hmm, what about the CI failures here? :thinking: 

As for the rest, I can merge + release once @weierophinney gives a :+1: on this

TotalWipeOut commented:
> Hmm, what about the CI failures here? thinking
> 
> As for the rest, I can merge + release once @weierophinney gives a +1 on this

thanks @Ocramius 

I need some help with the CI failures:
 - `Autocloser` check looks broken to me
 - PHPUnit tests for `latest` of 7.4, 8.0 and 8.1 all fail on `testValidatorChain`, but I can't see how my code effects that, and i can't seem to replicate it :/

I'm using latest `7.4.32`, and running `composer check`, but get no failures.

TotalWipeOut committed:
Fix Validator to work with laminas-validator 2.25+

Signed-off-by: Kevin Hamilton <kevin@eagleeye.com>

TotalWipeOut commented:
@Ocramius I only just realised earlier, that `latest` meant `composer update` the package, and not using `latest php patch version` :see_no_evil: sorry

From this, I could see that laminas-validator [2.25.0](https://github.com/laminas/laminas-validator/releases/tag/2.25.0) had broken the laminas-logs, Filter/Validator class, because ValidatorChain was now an iterator - which I have fixed! :tada: 

CI now passing, except the Autocloser, which looks broken?

Ocramius commented on src/Filter/Validator.php:
```diff
@@ -32,7 +32,7 @@ class Validator implements FilterInterface
     public function __construct($validator)
     {
-        if ($validator instanceof Traversable) {
+        if ($validator instanceof Traversable && ! $validator instanceof LaminasValidator) {
```
@gsteel this is an interesting case study of a scenario where **adding** an interface implementation to an existing system provokes a BC break :D

Ocramius commented on src/LoggerAbstractServiceFactory.php:
```diff
@@ -26,10 +26,18 @@ class LoggerAbstractServiceFactory implements AbstractFactoryInterface
+    }
 
     /**
-     * {@inheritDoc}
```
Not sure why these got changed, but no biggie

TotalWipeOut replied:
@Ocramius my editor suggested the change, because the interface docblock doesn't specify the exceptions - is it better  not make changes like this?

Ocramius replied:
It's not a problem, just noticed it and commented, but don't worry about it :)

Ocramius commented on src/LoggerAbstractServiceFactory.php:
```diff
@@ -26,10 +26,18 @@ class LoggerAbstractServiceFactory implements AbstractFactoryInterface
      * @var string
      */
-    protected $configKey = 'log';
+    protected $configKey;
```
Potentially problematic, if child classes override the constructor, but I don't think we should stress it: accessing state before the parent constructor is called would be a mistake anyway :)

Ocramius commented on src/LoggerAbstractServiceFactory.php:
```diff
@@ -63,21 +75,28 @@ public function __invoke(ContainerInterface $container, $requestedName, ?array $
     protected function getConfig(ContainerInterface $services)
     {
-        if ($this->config !== null) {
+        if (isset($this->config)) {
```
`isset()` should be reserved for cases where a property may not be declared: for nullable scenarios, please keep `!== null`, which leads to a notice or warning, if a property no longer exists for any reason.

Be very cautious around `isset()` and `empty()` usages: they are more complex than what they look like :)

TotalWipeOut replied:
@Ocramius ok, understood. The reason for the change was that I had typed that property as `protected array $config;` and this needed to change, as `!== null` caused a undefined notice.
I removed all the added typing, but thought that these props would eventually be typed, so left this here.

Ocramius replied:
Uninitialized properties after `__construct` are usually considered bugs :)

Ocramius commented on src/PsrLoggerAbstractAdapterFactory.php:
```diff
@@ -5,13 +5,18 @@
  * Allow to configure multiple loggers for application.
  */
-class PsrLoggerAbstractAdapterFactory extends LoggerAbstractServiceFactory
+class PsrLoggerAbstractAdapterFactory implements AbstractFactoryInterface
```
Potentially minor detail, but `protected processConfig()` no longer exists on this symbol now.

Anyway, OK with glancing over it

weierophinney commented on src/LoggerServiceFactory.php:
```diff
@@ -34,9 +44,108 @@ public function __invoke(ContainerInterface $container, $name, ?array $options =
+            $config['processor_plugin_manager'] = $services->get('LogProcessorManager');
+        }
+
+        if (! isset($config['writers'])) {
```
Based on how this is consumed in line 99, this should also include `|| ! is_iterable($config['writers'])` in the conditional to ensure we can actually traverse it.

TotalWipeOut replied:
I had to add this after the isset `if` aswell, due to the Logger itself only supporting real arrays
https://github.com/laminas/laminas-log/blob/2.15.x/src/Logger.php#L209

```php
        if (! is_array($config['writers'])) {
            $config['writers'] = iterator_to_array($config['writers']);
        }
```

weierophinney commented on src/LoggerServiceFactory.php:
```diff
@@ -34,9 +44,108 @@ public function __invoke(ContainerInterface $container, $name, ?array $options =
+
+        foreach ($config['writers'] as $index => $writerConfig) {
+            if (
+                isset($writerConfig['name'])
```
We're making the assumption that `$writerConfig` is an array or `ArrayAccess` implementation; should we assert before this that this is the case? It might help identify incorrect configuration earlier.

TotalWipeOut replied:
added guard statement

weierophinney commented on src/LoggerServiceFactory.php:
```diff
@@ -34,9 +44,108 @@ public function __invoke(ContainerInterface $container, $name, ?array $options =
+                // inject it into the configuration.
+                $manager                                         = $services->get($writerConfig['options']['manager']);
+                $config['writers'][$index]['options']['manager'] = $manager;
+            }
```
We should add a `continue` in the previous conditional, so that if we add additional conditions in this loop, we don't accidentally move on to them after this one matches.

weierophinney commented on test/LoggerServiceFactoryTest.php:
```diff
@@ -0,0 +1,272 @@
+use function count;
+use function extension_loaded;
+
+class LoggerServiceFactoryTest extends TestCase
```
I think we need a few additional tests:

- One for when `log.writers` is not iterable.
- One for when `log.writers[]` are not arrays/`ArrayAccess`

weierophinney reviewed (changes requested):
Sorry for taking so long to get back to you!

And thank you for stripping out the changes not related to the patch intention; makes it much clearer what you're accomplishing here.

I noted a few areas that need some additional changes; ironically, considering you were mainly refactoring, these were likely issues previously, but since you're refactoring, now would be a good time to address them.

TotalWipeOut committed:
Ensure writers config is valid, plus tests

Signed-off-by: Kevin Hamilton <kevin@eagleeye.com>

weierophinney reviewed (approved):
Wonderful work, @TotalWipeOut ! Thanks!
